// @slashauth/node-client/dist/organizations copy.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OrganizationController = void 0;
const query_1 = require("./query");
class OrganizationController {
    constructor(client_id, client_secret, apiClient) {
        this.client_id = client_id;
        this.client_secret = client_secret;
        this.apiClient = apiClient;
    }
    async hasOrgRole({ organizationID, address, role, }) {
        const encodedRole = Buffer.from(role, 'utf8').toString('base64');
        const urlParams = (0, query_1.signQuery)({
            input: {
                address,
                role: encodedRole,
                encoded: 'true',
            },
            secret: this.client_secret,
        });
        return this.apiClient.get(`/s/${this.client_id}/organizations/${organizationID}/has_role`, {
            queryParameters: {
                params: urlParams,
            },
        });
    }
    async getOrgMembershipsForUser({ userID, }) {
        const urlParams = (0, query_1.signQuery)({
            input: {
                userID,
            },
            secret: this.client_secret,
        });
        return this.apiClient.get(`/s/${this.client_id}/org_memberships`, {
            queryParameters: {
                params: urlParams,
            },
        });
    }
}
exports.OrganizationController = OrganizationController;
