// @slashauth/express/dist/middleware.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SlashauthMiddlewareExpress = void 0;
const token_1 = require("./utils/token");
class SlashauthMiddlewareExpress {
    constructor(client) {
        this.parseAuthToken = () => {
            return async (req, res, next) => {
                try {
                    const authHeader = req.headers.authorization || '';
                    const token = (0, token_1.parseToken)(authHeader);
                    const { data } = await this.client.user.validateToken({
                        token,
                    });
                    if (data) {
                        req.slashauth = {
                            wallet: data.wallet,
                            userID: data.userID,
                            isAuthed: true,
                            clientID: data.clientID,
                            getWalletAddress: data.getWalletAddress,
                        };
                    }
                }
                catch (err) { }
                next();
                return;
            };
        };
        this.hasRole = (role) => {
            return async (req, res, next) => {
                const authHeader = req.headers.authorization || '';
                const token = (0, token_1.parseToken)(authHeader);
                const { data } = await this.client.user.hasRoleToken({
                    role,
                    token,
                });
                if (data === null || data === void 0 ? void 0 : data.hasRole) {
                    next();
                    return;
                }
                return res.status(403).json({ error: `User does not meet role ${role}` });
            };
        };
        this.client = client;
    }
}
exports.SlashauthMiddlewareExpress = SlashauthMiddlewareExpress;
